File: cli/lib/context/status.js
'use strict';

var status = exports;

var Promise = require('es6-promise').Promise;

var Target = require('../context/target');
var output = require('../cli/output');

status.output = {};

status.output.failure = output.create(function () {
  console.log('Error determining status');
});

status.output.success = output.create(function (ctx, state) {
  var programName = ctx.program.name;

  console.log('Current Working Context:\n');
  console.log(
    'Identity: ' + state.user.body.name + ' (' + state.user.body.email + ')');
  console.log('Username: ' + state.user.body.username);

  if (state.target.disabled()) {
    return console.log('\nContext is disabled for your CLI,' +
      ' use \'' + programName + ' prefs\' to enabled it');
  }

  if (state.target.org === null) {
    return console.log('\nYou are not inside a linked working directory,' +
      ' use \'' + programName + ' link\' to link your project');
  }

  var service = state.target.service || '*';

  console.log('Org: ' + state.target.org);
  console.log('Project: ' + state.target.project);
  console.log('Environment: ' + state.target.environment);
  console.log('Service: ' + service);
  console.log('Instance: *');

  return console.log('\nCredential Path: /' + [
    state.target.org,
    state.target.project,
    state.target.environment,
    service,
    state.user.body.username,
    '*'
  ].join('/'));
});

status.execute = function (ctx) {
  return new Promise(function (resolve, reject) {
    if (!(ctx.target instanceof Target)) {
      throw new Error('Target must be on the context object');
    }

    return ctx.api.users.self().then(function (user) {
      if (!user) {
        throw new Error('Invalid response returned from the API');
      }

      resolve({
        user: user,
        target: ctx.target
      });
    }).catch(reject);
  });
};

File: cmd/status.go
package cmd

import (
	"fmt"
	"os"
	"strings"
	"text/tabwriter"

	"github.com/manifoldco/torus-cli/errs"
	"github.com/manifoldco/torus-cli/prefs"
	"github.com/manifoldco/torus-cli/ui"

	"github.com/urfave/cli"
)

func init() {
	status := cli.Command{
		Name:     "status",
		Usage:    "Show the current Torus status associated with your account and project",
		Category: "PROJECT STRUCTURE",
		Flags: []cli.Flag{
			stdEnvFlag,
			serviceFlag("Use this service.", "default", true),

			// These flags are hidden so we can still parse out the values
			// from the prefs files and env vars, but we don't display
			// them to the users in help.
			// A user could still set the flag on the command line though :(
			placeHolderStringFlag{
				StringFlag: cli.StringFlag{Name: "org", EnvVar: "TORUS_ORG", Hidden: true},
				Required:   true,
			},
			placeHolderStringFlag{
				StringFlag: cli.StringFlag{Name: "project", EnvVar: "TORUS_PROJECT", Hidden: true},
				Required:   true,
			},
			placeHolderStringFlag{
				StringFlag: cli.StringFlag{Name: "instance", EnvVar: "TORUS_INSTANCE", Value: "1", Hidden: true},
				Required:   true,
			},
		},
		Action: chain(
			ensureDaemon, ensureSession, loadDirPrefs, loadPrefDefaults,
			setUserEnv, statusCmd,
		),
	}

	Cmds = append(Cmds, status)
}

func statusCmd(ctx *cli.Context) error {
	preferences, err := prefs.NewPreferences()
	if err != nil {
		return err
	}

	if !preferences.Core.Context {
		msg := fmt.Sprintf("Context is disabled. Use '%s prefs' to enable it.", ctx.App.Name)
		return errs.NewExitError(msg)
	}

	err = checkRequiredFlags(ctx)
	if err != nil {
		fmt.Printf("You are not inside a linked working directory. "+
			"Use '%s link' to link your project.\n", ctx.App.Name)
		return nil
	}

	org := ctx.String("org")
	project := ctx.String("project")
	env := ctx.String("environment")
	service := ctx.String("service")

	w := tabwriter.NewWriter(os.Stdout, 2, 0, 1, ' ', 0)
	fmt.Fprintf(w, "%s:\t%s\n", ui.BoldString("Org"), org)
	fmt.Fprintf(w, "%s:\t%s\n", ui.BoldString("Project"), project)
	fmt.Fprintf(w, "%s:\t%s\n", ui.BoldString("Environment"), env)
	fmt.Fprintf(w, "%s:\t%s\n", ui.BoldString("Service"), service)
	w.Flush()

	parts := []string{"", org, project, env, service}
	credPath := strings.Join(parts, "/")
	fmt.Printf("\n%s: %s\n", ui.BoldString("Credential Path"), credPath)

	return nil
}
